api/src/main/java/org/pmiops/workbench/firecloud/FireCloudServiceImpl.java

```diff
@@ -47,18 +52,69 @@
   private static final String SAM_STATUS_NAME = "Sam";
   private static final String RAWLS_STATUS_NAME = "Rawls";
   private static final String GOOGLE_BUCKETS_STATUS_NAME = "GoogleBuckets";
+  private static final int MAX_ATTEMPTS = 3;
 
   @Autowired
   public FireCloudServiceImpl(Provider<WorkbenchConfig> configProvider,
       Provider<ProfileApi> profileApiProvider,
       Provider<BillingApi> billingApiProvider,
       Provider<GroupsApi> groupsApiProvider,
-      Provider<WorkspacesApi> workspacesApiProvider) {
+      Provider<WorkspacesApi> workspacesApiProvider,
+      Sleeper sleeper) {
     this.configProvider = configProvider;
     this.profileApiProvider = profileApiProvider;
     this.billingApiProvider = billingApiProvider;
     this.groupsApiProvider = groupsApiProvider;
     this.workspacesApiProvider = workspacesApiProvider;
+    this.sleeper = sleeper;
+  }
+
+  private static boolean isServiceUnavailableException(Exception e) {
+    if (e instanceof ApiException) {
+      int code = ((ApiException) e).getCode();
+      return (code > 500 && code <= 600);
+    }
+    return false;
+  }
+
+  private <T> T callFirecloudWithRetries(Callable<T> callable) throws ApiException {
```

Comment:
Why is this different than the ExceptionUtils version?

Reply:
ExceptionUtils uses an AbstractGoogleClientRequest (rather than Callable) and handles Google exceptions rather than Firecloud ones.

I wouldn't bother trying to write our own utility which abstracts further than this... might as well switch to a third party library if we are doing that.